crates/circuit/src/operations.rs

```diff
@@ -0,0 +1,811 @@
+        self.num_qubits()
+    }
+
+    #[getter]
+    pub fn get_num_clbits(&self) -> u32 {
+        self.num_clbits()
+    }
+
+    #[getter]
+    pub fn get_name(&self) -> &str {
+        self.name()
+    }
+}
+
+// This must be kept up-to-date with `StandardGate` when adding or removing
+// gates from the enum
+//
+// Remove this when std::mem::variant_count() is stabilized (see
+// https://github.com/rust-lang/rust/issues/73662 )
+pub const STANDARD_GATE_SIZE: usize = 18;
+
+impl Operation for StandardGate {
+    fn name(&self) -> &str {
+        match self {
+            Self::ZGate => "z",
+            Self::YGate => "y",
+            Self::XGate => "x",
+            Self::CZGate => "cz",
+            Self::CYGate => "cy",
+            Self::CXGate => "cx",
+            Self::CCXGate => "ccx",
+            Self::RXGate => "rx",
+            Self::RYGate => "ry",
+            Self::RZGate => "rz",
+            Self::ECRGate => "ecr",
+            Self::SwapGate => "swap",
+            Self::SXGate => "sx",
+            Self::GlobalPhaseGate => "global_phase",
+            Self::IGate => "id",
+            Self::HGate => "h",
+            Self::PhaseGate => "p",
+            Self::UGate => "u",
+        }
+    }
+
+    fn num_qubits(&self) -> u32 {
+        match self {
+            Self::ZGate => 1,
+            Self::YGate => 1,
+            Self::XGate => 1,
+            Self::CZGate => 2,
+            Self::CYGate => 2,
+            Self::CXGate => 2,
+            Self::CCXGate => 3,
+            Self::RXGate => 1,
+            Self::RYGate => 1,
+            Self::RZGate => 1,
+            Self::ECRGate => 2,
+            Self::SwapGate => 2,
+            Self::SXGate => 1,
+            Self::GlobalPhaseGate => 0,
+            Self::IGate => 1,
+            Self::HGate => 1,
+            Self::PhaseGate => 1,
+            Self::UGate => 1,
+        }
+    }
+
+    fn num_params(&self) -> u32 {
+        match self {
+            Self::ZGate => 0,
+            Self::YGate => 0,
+            Self::XGate => 0,
+            Self::CZGate => 0,
+            Self::CYGate => 0,
+            Self::CXGate => 0,
+            Self::CCXGate => 0,
+            Self::RXGate => 1,
+            Self::RYGate => 1,
+            Self::RZGate => 1,
+            Self::ECRGate => 0,
+            Self::SwapGate => 0,
+            Self::SXGate => 0,
+            Self::GlobalPhaseGate => 1,
+            Self::IGate => 0,
+            Self::HGate => 0,
+            Self::PhaseGate => 1,
+            Self::UGate => 3,
+        }
+    }
+
+    fn num_clbits(&self) -> u32 {
+        0
+    }
+
+    fn control_flow(&self) -> bool {
+        false
+    }
+
+    fn directive(&self) -> bool {
+        false
+    }
+
+    fn matrix(&self, params: &[Param]) -> Option<Array2<Complex64>> {
+        match self {
+            Self::ZGate => match params {
+                [] => Some(aview2(&gate_matrix::ZGATE).to_owned()),
+                _ => None,
+            },
```

Comment:
I can make it `Result<Option<...>>`, since I think the no matrix case isn't an error and we want the trait to be able to distinguish that from an error case where the number of parameters doesn't match the operation's expectation.